### ISSUE TITLE
Generic Anchor/Link

### ISSUE THREAD

**Issue by hepiyellow** [wix/wix-style-react#2544]
## Use Cases

- Non-Focusable - (Avatar, any presentational component)
- Focusable - only Button Family (`<button/>`)

## Button use case

- [ ] External `<Link/>` wrapper
- Button as Link

### External `<Link/>` wrapper (ReactRouter)

This:
```
<Link href="" aria-label="Link to blah" className={buttoonStyles.focusParent}>
   <Button onClick={undefined}/>
</Link>
```
renders this:
```
<a>
   <button>
</a>
```

#### Requirements
- Focus
  - Button without `onClick` should no be focusable
  - Focus-Ring

#### Solution 1
```js
<Link href="" aria-label="Link to blah" className={wsrLinkStyles.linkWrapper}>
   <Button onClick={undefined}/>
</Link>
```

> Button.st.css
```css
.linkWrapper:focus-ring > .foo {
  color:red;
}
```

#### Solution 2a- comp prop
```js
<Button comp={<Link href="http"/>}>
<Button comp={<a href="http"/>}>
```
Instead of 
```js
<button class="root">
    Hello
</button
```
Becomes (replacing):
```js
<a >
  Hello
</a
```
OR
Becomes (wrapping):
```js
<a >
  <button>
     Hello
  <button>
</a
```
- Issue: `<button/>` MUST be non-focusable (Can we hack it?)

##### CONS
DEV: issue: TS types propTypes for HTMLButtonElement are no correct if root element is NOT a `<button/>`
tagName transition can be:
button -> a
div -> a 
span -> a

> If we define `display, padding, box-sizing` on class `root` the
PROD: 
Styling

#### Solution 2b - hide the linkWrapper class
```js
<Button comp={<Link href="http"/>}>
```
Becomes:
```js
<Link class="prev_class wsrFocusableParent">
  <Button/> // not focusable
</Link>
```

#### API option 1 - Union
```js
interface LinkButtonProps extends HTMLAnchorAttributes, HTMLButtonAttributes {
}
<LinkButton
   href="http:/..."
   size={large}
>
```

#### CONS
Need one for every Button family (`<LinkIconButton/>`)...
But if 

#### API option 2- Consumer wraps with `<a>`
```js
<a href="" aria-label="Link to blah" className={buttoonStyles.focusParent}>
   <Button onClick={undefined}/>
</a>
```

##### CONS
- `buttoonStyles.focusParent` class 

#### API option 3 - Core LinkButton
Use same "theming" (variaion group) like ButtonNext.

```js
<LinkButton className={iconButton()} />
``` 

#### API option 4 - core `<Button href/>`
```js
<Button href=""/>
```
Will render <a> instead of <button> inside.

PropsInterface
```js
interface ButtonProps extends HTMLAnchorAttributes, HTMLButtonAttributes {
}
```

CONS: button has `type` which is invalid for anchor

  

## Conclusions

TL;DR

```js
<Button href="http"/>

<RouterLink href className={wsr.linkWrapper}>
   <Avatar/>
<RouterLink
```

**Comment by shlomitc** [comment]
clickable components that need to be rendered with a different wrapper will get an `as` prop that will render the component
For example, see the Button PR: https://github.com/wix/wix-ui/pull/1030

@hepiyellow - feel free to add this to the contribution guide

**Comment by hepiyellow** [comment]
@shlomitc @mykas 

Just some thoughts...

There are 2 concerns when changing a component's root element (or wrapping):
1. Styling: Focus styles, Overriding default browser styles
1. Behavior: avoiding nested focusable elements. (e.g. `<a><button/></a>`) 

### Button solution (Clickable native elements)

With the `<Button as={'a'} href="http://"/>` solution, we solve both concerns. 

And we couldn't find any other clickable native elements which might need to become links (e.g. `<input/>` will never be a link)

### Non-Clickable native elements (VisualComponents)

I think that components which have a non-clickable root element (`<div/>`) should not be clickable (no `onClick` prop`.

We don't know if the consumer wants it to be `<button/>` or a `<a/>`.
The consumer can always wrap it with a `<button/>` or `<a/>` (the behavior concern is not relevant).

#### Styling concern

##### Focus Styles

In both cases (wrapper or `as`), we don't know what is the new root element native type (Since the wrapper can be a Custom element (e.g. we don't know that `<Link/>` renders ` <a/>`).

So :
- We don't know if the new root element is focusable.
- We don't know if the new root element is, for example,  an `<a/>` so we don't know we need to reset the default browser `decoration: underline`.

#### Solution with `as`, `href`, `onClick`, `tabIndex` props

Allowing to pass `href` or `onClick` to tells us something about the new root element.
- If an `href` is passed, we can assume the new root element is an `<a/>` and disable `decorator: underline`.
- If `href` or `onClick` or `tabIndex` is passed, we can assume the new root element is focusable, and apply focus styles. (In this case, a `focus-visible` polyfill would be a better solution than our FocusableHOC).

> Instead of adding these props to each VisualComponent, we can have a HOC ( I think) that will make any VisualComponent `As-able` (or `Mutable` or `Wrappable` (-: )

### Conclusion

So I think we need to propose a fully working solution of all use-cases.
